src/python/pants/core/goals/init.py

```diff
@@ -0,0 +1,224 @@
+    help = "Generate config for working with Pants."
+
+
+class Init(Goal):
+    subsystem_cls = InitSubsystem
+
+
+@dataclass(frozen=True)
+class EditBuildFilesRequest:
+    putative_targets: PutativeTargets
+
+
+@dataclass(frozen=True)
+class EditedBuildFiles:
+    digest: Digest
+    created_paths: Tuple[str, ...]
+    updated_paths: Tuple[str, ...]
+
+
+def group_by_build_file(ptgts: Iterable[PutativeTarget]) -> Dict[str, List[PutativeTarget]]:
+    ret = defaultdict(list)
+    for ptgt in ptgts:
+        ret[ptgt.build_file_path].append(ptgt)
+    return ret
+
+
+@rule
+async def edit_build_files(req: EditBuildFilesRequest) -> EditedBuildFiles:
+    ptgts_by_build_file = group_by_build_file(req.putative_targets)
+    existing_build_files_contents = await Get(DigestContents, PathGlobs(ptgts_by_build_file.keys()))
+    existing_build_files_contents_by_path = {
+        ebfc.path: ebfc.content for ebfc in existing_build_files_contents
+    }
+
+    def make_content(bf_path: str, pts: Iterable[PutativeTarget]) -> FileContent:
+        existing_content = existing_build_files_contents_by_path.get(bf_path)
+        new_content = ([] if existing_content is None else [existing_content.decode()]) + [
+            pt.generate_build_file_stanza() for pt in pts
+        ]
+        new_content_bytes = "\n\n".join(new_content).encode()
+        return FileContent(bf_path, new_content_bytes)
+
+    new_digest = await Get(
+        Digest,
+        CreateDigest([make_content(path, ptgts) for path, ptgts in ptgts_by_build_file.items()]),
+    )
+
+    updated = set(existing_build_files_contents_by_path.keys())
+    created = set(ptgts_by_build_file.keys()) - updated
+    return EditedBuildFiles(new_digest, tuple(sorted(created)), tuple(sorted(updated)))
+
+
+@goal_rule
+async def init(
+    init_subsystem: InitSubsystem,
+    console: Console,
+    workspace: Workspace,
+    union_membership: UnionMembership,
+) -> Init:
+    putative_target_request_types = union_membership[PutativeTargetsRequest]
+    putative_target_reqs = [req_type() for req_type in putative_target_request_types]
+    putative_targets_results = await MultiGet(
+        Get(PutativeTargets, PutativeTargetsRequest, req) for req in putative_target_reqs
+    )
+    all_putative_targets = PutativeTargets.merge(putative_targets_results)
+    all_existing_tgts = await Get(Targets, AddressSpecs([DescendantAddresses("")]))
+    addr_to_existing_tgt = {
+        f"{tgt.address.spec_path}:{tgt.address.target_name}": tgt for tgt in all_existing_tgts
+    }
+
+    conflicting_ptgts: List[PutativeTarget] = []
+    nonconflicting_ptgts: List[PutativeTarget] = []
+
+    for ptgt in all_putative_targets:
+        (
+            conflicting_ptgts if ptgt.address_spec in addr_to_existing_tgt else nonconflicting_ptgts
+        ).append(ptgt)
+
+    conflicting_ptgts_by_build_file = group_by_build_file(conflicting_ptgts)
+    nonconflicting_ptgts_by_build_file = group_by_build_file(nonconflicting_ptgts)
```

Comment:
This conflict detection should be working to prevent multiple targets owning a file, right? Does something about the generated address name prevent that? It seems like maybe this merge should be happening based on owned file names instead?

It also seems like this merge should be preventing conflicts with existing targets based on overlapping globs, I think. We know that no existing target owns a file owned by one of the `PutativeTarget`s (assuming the implementations are correct), but it looks possible currently for a `PutativeTarget` using default globs to create a target that owns a file already owned by another target.

Reply:
Hmm, right now this doesn't prevent that. The conflicts are based on target address. We should probably check for overlapping globs too.

Reply:
+1. Things majorly break in dep inference world if >1 target owns the same source file. It's key that this goal never introduces such a conflict.